### ISSUE TITLE
OTS should be able to be built using system dependencies

### ISSUE THREAD

**Issue by MicroJoe** [khaledhosny/ots#221]
This library should be able to be built from system dependencies like `libbrotli-dev`, `liblz4-dev` and `libwoff-dev` instead of only allowing to build from the copied `third_party/` submodules.

Possibility to build from the `third_party/` submodules should still be kept if the system dependencies cannot be found on the system.

(note to maintainer: creating this bug for tracking the issue in public, separate from any PR proposal)

**Comment by musicinmybrain** [comment]
Thanks for filing this issue and offering a PR! I was working on the same thing today; I am working on packaging OTS for Fedora.

I have an alternative PR that uses the external/system libraries only when explicitly requested via meson options, otherwise maintaining the current behavior of always using the bundled/subproject/submodule copies. I have no preference between your approach and mine—whichever the upstream developers prefer is fine by me. I do have a few notes that I think might be helpful:

- Nothing uses libbrotlidec; libbrotlienc is a transitive dependency of libwoff2, so it can be omitted too, as pkgconfig will be aware of this; and the same for libbrotlicommon, which is a transitive dependency of libwoff2enc.
- Nothing uses libwoff2dec; libwoff2common is a transitive dependency of libwoff2dec.
- Summary of the above: of lib{brotli,woff2}{enc,dec,common}, you only need libwoff2dec as an explicit dependency when using pkgconfig.
- Your PR missed handling liblz4.
- It would be best to handle gtest as well; while the gtest developers do not advocate using it as an external library, it works well in the context of a Linux distribution, and most distros will patch the build to use an external gtest anyway.

Here is my alternative PR: #223

**Comment by MicroJoe** [comment]
Hello, what a coincidence! I am myself packaging this for Debian. ;)

> I have an alternative PR that uses the external/system libraries only when explicitly requested via meson options, otherwise maintaining the current behavior of always using the bundled/subproject/submodule copies. I have no preference between your approach and mine—whichever the upstream developers prefer is fine by me.

I prefer your approach, because I was going to explicitly remove the if/else I introduced in the Debian package to be sure that nothing gets downloaded from the Internet.

> Summary of the above: of lib{brotli,woff2}{enc,dec,common}, you only need libwoff2dec as an explicit dependency when using pkgconfig.

Good to know, I really did not dug into this and linked everything.

> Your PR missed handling liblz4.

Whoops, correct.

> It would be best to handle gtest as well; while the gtest developers do not advocate using it as an external library, it works well in the context of a Linux distribution, and most distros will patch the build to use an external gtest anyway.

I was going to add this as well, but we currently have a bug in Debian that makes the `googletest` package not working with meson. As soon as this bug is fixed in Debian I think using the `googletest` package as a dependency should be easy.

Overall I really prefer your approach in #223 with the use of meson options. I hope upstream chooses it over my hack. I will close my PR so that it does not get accidentally merged if upstream does not see this discussion.

**Comment by musicinmybrain** [comment]
Working on a new PR based on feedback.

**Comment by musicinmybrain** [comment]
So, it looks like the Appveyor CI failed for my draft PR, because:

```
Couldn't use fallback subproject in subprojects\lz4 for the dependency liblz4 
Reason: subprojects\lz4-1.9.3\contrib/meson/meson\meson.build:12: Tried to access unknown option "lz4:c_std".
```

followed by trying and failing to use pkg-config to find a system copy.

Here’s the trivial relevant line from `subprojects/lz4-1.9.3/contrib/meson/meson/meson.build`:

```
c_std = get_option('c_std')
```

The same pinned versions `meson==0.49.2 ninja==1.9.0` work on my Linux system using the fallback subproject. Has anyone ever seen anything like this? The zlib and gtest wraps are working just fine. I don’t have access to a Windows installation to debug this, so the best I can do is push things to my PR and see what the CI does.

**Comment by musicinmybrain** [comment]
The draft PR is now passing CI with everything but woff2 converted to meson subprojects. The following accommodations were required:

- No more unity build support, because brotli uses static symbols with the same names in different translation units; therefore, I disabled unity builds in Travis CI.
- Minimum meson version is now 0.51 across the board; this flows up from brotli, which requires 0.50, plus https://github.com/mesonbuild/meson/issues/5070, which is fixed in 0.51. (Travis CI was already using 0.52.)
- Changed meson version on AppVeyor to 0.55.2; this is the earliest version that fixes the lz4 subproject bug I mentioned in my previous comment. I do not know what particular meson change or bug fix is responsible. It still appears to be Windows-specific.

Next up is a wrap for woff2, hopefully.

**Comment by khaledhosny** [comment]
>     * Changed meson version on AppVeyor to 0.55.2;

I’m fine with raising to the required meson version (the only reason CI builds are pinned to a specific version is to avoid unexpected build breakages that sometimes happen with newer meson versions, but otherwise I don’t mind requiring even the latest version if needed).
